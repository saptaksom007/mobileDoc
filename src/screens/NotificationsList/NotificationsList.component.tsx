import React, { Component } from 'react'

import { View, InteractionManager } from 'react-native'
import { Icon } from 'react-native-elements'
import TouchableNativeFeedbackSafe from '@expo/react-native-touchable-native-feedback-safe'
import { connect } from 'react-redux'
import EmptyList from 'components/EmptyList/EmptyList.component'
import ListContainer from 'components/ListContainer/ListContainer.component'
import i18n from 'ex-react-native-i18n'
import { decodeText } from 'utilities/html'
import { compose } from 'recompose'
import { propOr } from 'ramda'
import { Actions as NavAction } from 'navigation/SagaNavigation'
import filters from 'common-docdok/lib/utils/filters'
import Conversation from 'components/Conversation/Conversation.component'
import withSubtitle from 'screens/enhancers/withSubtitle'
import withoutKeyboard from 'screens/enhancers/withoutKeyboard'
import dateUtil from 'common-docdok/lib/utils/dateUtils'
import { messagingActions } from 'common-docdok/lib/domain/messaging/actions'
import { withLoader } from 'screens/enhancers/withLoader'
import { getTestId } from 'utilities/environment'
import { Color } from 'constants/Color'
import { getSurveyUrlById } from 'utilities/survey'

import styles from './NotificationsList.styles'
import { Action } from 'redux'

const testIdNotificationsList = getTestId('notifications-list')

interface Props {
  notificationsList: any[]
  dispatch(action: Action): void
  surveyDefinitionsById: { [id: number]: any }
  users: any
  notificationId: number
  appointments: {
    [id: number]: any & { physicianAvatarPicture?: string }
  }
}

export class NotificationsList extends Component<Props> {
  componentDidMount() {
    InteractionManager.runAfterInteractions(() => {
      const { dispatch, notificationId } = this.props
      dispatch(messagingActions.loadMessagesRequested(notificationId))
    })
  }

  getAppointmentTitle = (
    appointmentType: 'Appointment' | 'Schedule' | 'Confirmation'
  ) => {
    switch (appointmentType) {
      case 'Appointment': {
        return i18n.t('appointment.subtitle')
      }
      case 'Schedule': {
        return i18n.t('scheduleAppointment.subtitle')
      }
      case 'Confirmation': {
        return i18n.t('appointmentConfirmation.subtitle')
      }
      default: {
        return i18n.t('appointment.subtitle')
      }
    }
  }

  dissmissAll = () => {
    const { dispatch, notificationsList } = this.props
    dispatch(messagingActions.markAsCheckedRequested(notificationsList, true))
  }

  selectNotification(notification: any) {
    const { dispatch } = this.props
    switch (notification.meta.notificationtype) {
      case 'SURVEY': {
        const url = getSurveyUrlById(notification.meta.id)
        dispatch(NavAction.push('surveywebview', { url }))
        break
      }

      case 'APPOINTMENT': {
        const appointmentId = notification.meta.appointmentId
        const { appointmentType } = notification.meta
        switch (appointmentType) {
          case 'Appointment': {
            dispatch(NavAction.push('appointment', { appointmentId }))
            return
          }
          case 'Schedule': {
            dispatch(NavAction.push('scheduleappointment', { appointmentId }))
            return
          }
          case 'Confirmation': {
            dispatch(
              NavAction.push('appointmentconfirmation', { appointmentId })
            )
            return
          }
          default:
            dispatch(NavAction.showLocalWarning(i18n.t('alert.notImplemented')))
            return
        }
      }

      default: {
        dispatch(NavAction.showLocalWarning(i18n.t('alert.notImplemented')))
      }
    }
  }

  renderSurveyNotification(notification: any) {
    const { users } = this.props
    const user = users[notification.userRef]
    const avatarUrl = user && user.avatarPicture

    const surveyName =
      notification.meta.surveyName || decodeText(notification.text)

    const content =
      notification.meta.subtype === 'SURVEY_REMINDER'
        ? i18n.t('notificationslist.surveyReminder', {
          content: surveyName
        })
        : i18n.t('notificationslist.survey', {
          content: surveyName
        })
    return (
      <Conversation
        iconName="show-chart"
        avatarUrl={user && avatarUrl}
        title={user && `${user.firstName} ${user.lastName}`}
        detailNumberOfLines={5}
        content={content}
        postedAt={dateUtil.toDate(notification.postedAt)}
        onPress={() => this.selectNotification(notification)}
      />
    )
  }

  renderAppointmentNotification(notification: any) {
    const { appointments } = this.props
    const appointment = appointments[notification.meta.appointmentId]
    if (!appointment) {
      return null
    }
    const { appointmentType } = notification.meta
    return (
      <Conversation
        iconName="event"
        avatarUrl={appointment && appointment.physicianAvatarPicture}
        title={this.getAppointmentTitle(appointmentType)}
        detailNumberOfLines={2}
        content={notification.text}
        postedAt={dateUtil.toDate(notification.postedAt)}
        onPress={() => this.selectNotification(notification)}
      />
    )
  }

  renderDefault(notification: any) {
    const { users } = this.props
    const user = users[notification.userRef]
    const avatarUrl = user && user.avatarPicture
    const firstName = user && user.firstName
    const lastName = user && user.lastName

    return (
      <Conversation
        iconName="notifications"
        avatarUrl={avatarUrl}
        title={`${firstName || ''} ${lastName || ''}`}
        content={notification.text && decodeText(notification.text)}
        postedAt={dateUtil.toDate(notification.postedAt)}
        onPress={() => this.selectNotification(notification)}
      />
    )
  }

  renderRow = ({ item: rowData }: any) => {
    switch (rowData.meta.notificationtype) {
      case 'SURVEY': {
        return this.renderSurveyNotification(rowData)
      }

      case 'APPOINTMENT': {
        return this.renderAppointmentNotification(rowData)
      }

      default:
        return this.renderDefault(rowData)
    }
  }

  render() {
    const { dispatch, notificationsList } = this.props

    if (!notificationsList || notificationsList.length === 0) {
      return (
        <EmptyList
          testID={testIdNotificationsList}
          text={i18n.t('notificationslist.empty')}
          show
          onPress={() => dispatch(messagingActions.getSubscriptionsRequested())}
        />
      )
    }

    return (
      <View style={styles.container}>
        <View style={styles.button}>
          <TouchableNativeFeedbackSafe
            style={styles.dismiss}
            onPress={this.dissmissAll}
          >
            <Icon name="delete" color={Color.darkText} />
          </TouchableNativeFeedbackSafe>
        </View>

        <ListContainer
          testID={testIdNotificationsList}
          data={notificationsList}
          renderItem={this.renderRow}
        />
      </View>
    )
  }
}

function select({
  survey: { surveyDefinitionsById },
  userCache: { users },
  messaging: { conversations, checkedMessages },
  appointmentslist: { rawList: appointments }
}: any) {
  const notificationConv = filters
    .values(conversations)
    .filter((conv: any) => conv.type === 'NOTIFICATION')[0]

  const notificationId: string = propOr(0, 'id', notificationConv)
  const messages: any[] = propOr([], 'messages', notificationConv)
  const latestCheckedMessage: any[] = propOr(0, notificationId, checkedMessages)

  const notificationsList = messages
    .filter(
      (m: any) =>
        latestCheckedMessage.length === 0 || m.sequenceNo > latestCheckedMessage
    )
    .sort((m1: any, m2: any) => m2.sequenceNo - m1.sequenceNo)

  return {
    notificationsList,
    surveyDefinitionsById,
    users,
    notificationId,
    appointments:
      appointments &&
      appointments.reduce(
        (prev: any, curr: any) => ({ ...prev, [curr.id]: curr }),
        {}
      )
  }
}

export default compose<Props, any>(
  withoutKeyboard,
  connect(select),
  withLoader,
  withSubtitle('notificationslist.subtitle')
)(NotificationsList)
