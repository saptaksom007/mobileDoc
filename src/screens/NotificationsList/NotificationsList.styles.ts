import { StyleSheet } from 'react-native'
import { Color } from 'constants/Color'

export default StyleSheet.create({
  container: {
    flex: 1
  },
  button: {
    position: 'absolute',
    bottom: 25,
    left: 0,
    right: 0,
    justifyContent: 'center',
    alignItems: 'center',
    zIndex: 1000
  },
  dismiss: {
    backgroundColor: Color.borderColor,
    width: 60,
    height: 60,
    borderRadius: 30,
    justifyContent: 'center',
    alignItems: 'center'
  }
})
