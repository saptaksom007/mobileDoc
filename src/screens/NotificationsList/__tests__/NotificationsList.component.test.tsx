import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import { NotificationsList } from '../NotificationsList.component'

jest.mock(
  'components/Conversation/Conversation.component',
  () => 'Conversation'
)

it('<NotificationsList/> render correctly', () => {
  const tree = renderer
    .create(
      <NotificationsList
        dispatch={() => {}}
        notificationsList={[]}
        surveyDefinitionsById={{ 123: {} }}
        notificationId={123}
        appointments={{
          15: {}
        }}
        users={{
          '0cbe4673-8eec-48ad-adad-594f3c02e729': {
            avatarPicture: ''
          }
        }}
      />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('<NotificationsList/> render correctly with survey', () => {
  const tree = renderer
    .create(
      <NotificationsList
        surveyDefinitionsById={{ 123: {} }}
        notificationId={123}
        appointments={{
          15: {}
        }}
        dispatch={() => {}}
        users={{
          '0cbe4673-8eec-48ad-adad-594f3c02e729': {
            avatarPicture: ''
          }
        }}
        notificationsList={[
          {
            id: 1443,
            uuid: '6f0e507d-e9fe-478a-ae02-55006c4745fb',
            text: '',
            userRef: '0cbe4673-8eec-48ad-adad-594f3c02e729',
            postedAt: '2018-07-05T10:19:07+02:00',
            sequenceNo: 1,
            mediaResource: null,
            conversationId: 650,
            meta: {
              surveyId: '15',
              patientuuid: 'PAT-251b7fdc-2eec-48fa-ae80-34a725e669c3',
              surveyName: 'Erstanmeldung Erwachsene',
              senderRef: '0cbe4673-8eec-48ad-adad-594f3c02e729',
              subtype: 'SURVEY_COMPLETED',
              notificationtype: 'SURVEY',
              id: '18',
              creationDate: '2018-07-05T10:19:02+02:00',
              url:
                'https://qa.dev.docdok.ch/limesurvey/index.php/16953?token=a8FoImzZhmoVzt0&newtest=Y&lang='
            }
          }
        ]}
      />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('<NotificationsList/> render correctly with appointment', () => {
  const tree = renderer
    .create(
      <NotificationsList
        surveyDefinitionsById={{ 123: {} }}
        notificationId={123}
        dispatch={() => {}}
        appointments={{
          15: {}
        }}
        users={{
          '0cbe4673-8eec-48ad-adad-594f3c02e729': {
            avatarPicture: ''
          }
        }}
        notificationsList={[
          {
            id: 1443,
            uuid: '6f0e507d-e9fe-478a-ae02-55006c4745fb',
            text: '',
            userRef: '0cbe4673-8eec-48ad-adad-594f3c02e729',
            postedAt: '2018-07-05T10:19:07+02:00',
            sequenceNo: 1,
            mediaResource: null,
            conversationId: 650,
            meta: {
              appointmentId: '15',
              notificationtype: 'APPOINTMENT'
            }
          }
        ]}
      />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
