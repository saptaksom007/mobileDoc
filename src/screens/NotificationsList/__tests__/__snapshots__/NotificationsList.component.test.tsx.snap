// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotificationsList/> render correctly 1`] = `
<View
  accessibilityLabel="notifications-list"
  style={
    Object {
      "alignItems": "center",
      "height": 1294,
      "justifyContent": "center",
    }
  }
  testID="notifications-list"
>
  <Text
    allowFontScaling={false}
    ellipsizeMode="tail"
    onPress={[Function]}
    style={
      Array [
        Object {
          "fontFamily": "TitilliumWeb-Regular",
          "fontSize": 17.5,
          "lineHeight": 22,
          "overflow": "visible",
        },
        Object {
          "color": "rgba(27, 106, 158, 0.85)",
          "fontSize": 17.5,
          "lineHeight": 21,
        },
        Object {
          "color": "#999",
          "marginTop": -40,
        },
      ]
    }
  >
    notificationslist.empty
  </Text>
</View>
`;

exports[`<NotificationsList/> render correctly with appointment 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "bottom": 25,
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "zIndex": 1000,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#E7ECEE",
            "borderRadius": 30,
            "height": 60,
            "justifyContent": "center",
            "width": 60,
          }
        }
      />
    </View>
  </View>
  <View
    accessibilityLabel="notifications-list"
    style={
      Object {
        "backgroundColor": "#fff",
        "flex": 1,
      }
    }
    testID="notifications-list"
  >
    <RCTScrollView
      data={
        Array [
          Object {
            "conversationId": 650,
            "id": 1443,
            "mediaResource": null,
            "meta": Object {
              "appointmentId": "15",
              "notificationtype": "APPOINTMENT",
            },
            "postedAt": "2018-07-05T10:19:07+02:00",
            "sequenceNo": 1,
            "text": "",
            "userRef": "0cbe4673-8eec-48ad-adad-594f3c02e729",
            "uuid": "6f0e507d-e9fe-478a-ae02-55006c4745fb",
          },
        ]
      }
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      style={
        Object {
          "backgroundColor": "#fff",
          "flex": 1,
        }
      }
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <Conversation
            content=""
            detailNumberOfLines={2}
            iconName="event"
            onPress={[Function]}
            postedAt={2018-07-05T08:19:07.000Z}
            title="appointment.subtitle"
          />
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;

exports[`<NotificationsList/> render correctly with survey 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "bottom": 25,
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "zIndex": 1000,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#E7ECEE",
            "borderRadius": 30,
            "height": 60,
            "justifyContent": "center",
            "width": 60,
          }
        }
      />
    </View>
  </View>
  <View
    accessibilityLabel="notifications-list"
    style={
      Object {
        "backgroundColor": "#fff",
        "flex": 1,
      }
    }
    testID="notifications-list"
  >
    <RCTScrollView
      data={
        Array [
          Object {
            "conversationId": 650,
            "id": 1443,
            "mediaResource": null,
            "meta": Object {
              "creationDate": "2018-07-05T10:19:02+02:00",
              "id": "18",
              "notificationtype": "SURVEY",
              "patientuuid": "PAT-251b7fdc-2eec-48fa-ae80-34a725e669c3",
              "senderRef": "0cbe4673-8eec-48ad-adad-594f3c02e729",
              "subtype": "SURVEY_COMPLETED",
              "surveyId": "15",
              "surveyName": "Erstanmeldung Erwachsene",
              "url": "https://qa.dev.docdok.ch/limesurvey/index.php/16953?token=a8FoImzZhmoVzt0&newtest=Y&lang=",
            },
            "postedAt": "2018-07-05T10:19:07+02:00",
            "sequenceNo": 1,
            "text": "",
            "userRef": "0cbe4673-8eec-48ad-adad-594f3c02e729",
            "uuid": "6f0e507d-e9fe-478a-ae02-55006c4745fb",
          },
        ]
      }
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      style={
        Object {
          "backgroundColor": "#fff",
          "flex": 1,
        }
      }
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <Conversation
            avatarUrl=""
            content="notificationslist.survey"
            detailNumberOfLines={5}
            iconName="show-chart"
            onPress={[Function]}
            postedAt={2018-07-05T08:19:07.000Z}
            title="undefined undefined"
          />
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
